### PR TITLE
ENH: nd state support, fzp states, plc example motion object

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

- Add support for multidimensional states while maintaining backwards compatibility with singledimensional states.
- Add states to TMO's FZP spectrometer.
- Add some example devices

## Motivation and Context

https://github.com/pcdshub/lcls-twincat-motion/pull/163

There are now multidimensional states in lcls-twincat-motion
This matches the PV structure

The expected API for defining a multidimensional state is a natural extension for the single dimensional states:
- subclass the states class
- update the config component with the motor count and the state count

Behind the scenes, this is messing with the `__new__` constructor to dynamically make motor-and-state-count specific subclasses as needed.

Here's how I implemented the FZP states class in this PR as an example of how a user would use this:
```
class FZPStates(TwinCATStatePMPS):
    """
    FZP ND States Setup
    """
    config = UpCpt(state_count=15, motor_count=3)
```

## How Has This Been Tested?

- Interactively by popping up the typhos screen for the FZP (and the cli version using happi load)
- Slightly extended unit tests

## Where Has This Been Documented?

I guess I'll put it in the release notes

## Screenshots (if appropriate):
image

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by ZLLentz** [comment]
Works with typhos with my test ioc right now:

image

File: pcdsdevices/spectrometer.py [hunk 1]
```diff
@@ -307,6 +309,13 @@ def calc_lightpath_state(self, **kwargs) -> LightpathState:
         )
 
 
+class FZPStates(TwinCATStatePMPS):
+    """
+    FZP ND States Setup
```
Comment (klauer) [inline pcdsdevices/spectrometer.py:314]: I know this is fresnel zone plate, but does everyone else? ~Macro~ acronym expansion in docstring 🙏

File: docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+-----------
+- N/A
+
+Features
+--------
+- ND TwinCAT states are now supported.
+
+Device Updates
+--------------
+- Included the FZP 3D states on the TMOSpectrometer device.
```
Comment (klauer) [inline docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst:14]: Same comment about acronyms

File: docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+New Devices
+-----------
+- Included example devices and components that correspond to
+  lcls-plc-example-motion.
```
Comment (klauer) [inline docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst:19]:
```suggestion
  `lcls-plc-example-motion <https://github.com/pcdshub/lcls-plc-example-motion>`_.
```
(I think I have the syntax right there...)

File: pcdsdevices/example.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    siml2l = Cpt(ExampleL2L, "L2L:")
+
+    def __init__(self, prefix: str = "PLC:TST:MOT:", name: str = "plc_example_motion"):
+        super().__init__(prefix, name=name)
+
+    def all_pvnames(self) -> list[str]:
+        """
+        Get all the pvnames that should be included in the IOC.
+
+        This is to help with debug.
```
Comment (klauer) [inline pcdsdevices/example.py:105]: Nit
```suggestion
        This is to help with debugging.
```

File: pcdsdevices/example.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        return list(
+            walk.item.pvname for walk in self.walk_signals() if hasattr(walk.item, "pvname")
+        )
+
+    def disconnected_pvnames(self) -> list[str]:
+        """
+        Get all of the pvnames that are currently disconnected.
+
+        This is to help with debug.
```
Comment (klauer) [inline pcdsdevices/example.py:115]: Pick
```suggestion
        This is to help with debugging.
```

File: pcdsdevices/state.py [hunk 6]
```diff
@@ -593,31 +621,88 @@ class FakeTwinCATStateConfigDynamic(TwinCATStateConfigDynamic):
...
+    else:
+        return f"m{motor_index}_{state_attr}"
+
+
+def state_config_dotted_attribute(
+    config_attr: str,
+    state_count: int,
+    motor_count: int = 1,
+    first_motor_only: bool = False,
+) -> list[str | None]:
```
Comment (klauer) [inline pcdsdevices/state.py:665]: It looks like this is always `list[str]` now?

File: pcdsdevices/state.py [hunk 7]
```diff
@@ -629,19 +714,27 @@ def state_config_dotted_names(state_count: int) -> list[str | None]:
...
+        motor_count=motor_count,
+        first_motor_only=True,
+    )
 
 
-def state_config_dotted_velos(state_count: int) -> list[str | None]:
+def state_config_dotted_velos(
+    state_count: int,
+    motor_count: int = 1,
+) -> list[str | None]:
```
Comment (klauer) [inline pcdsdevices/state.py:737]: Also just `list[str]`, I think?

File: docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst [hunk 8]
```diff
@@ -0,0 +1,31 @@
+1136 enh_nd_twincat_states
+##########################
+
+API Changes
+-----------
+- N/A
+
+Features
+--------
+- ND TwinCAT states are now supported.
```
Comment (klauer) [inline docs/source/upcoming_release_notes/1136-enh_nd_twincat_states.rst:10]:
```suggestion
- ND (N-dimensional) TwinCAT states are now supported.
```

**Review by klauer — APPROVED** [review]
* Some nits and picks as per usual
* It's really nice to have the example class devices built-in
* Overall LGTM 👍

File: pcdsdevices/state.py (hunk 7 above)
Comment (tangkong) [reply pcdsdevices/state.py:737]: in line 726 there's always `None` in the list at index 0.  I do wonder why this is necessary though..  (I'm sure it works it just seems weird)

File: pcdsdevices/state.py (hunk 6 above)
Comment (tangkong) [reply pcdsdevices/state.py:665]: This one does seem like it's only `list[str]`

**Review by tangkong — APPROVED** [review]
Looks good to me, I think expanding the examples file is going to be good moving forward.  Seeing tests always puts my mind at ease

File: pcdsdevices/state.py (hunk 7 above)
Comment (klauer) [reply pcdsdevices/state.py:737]: The diff makes this tougher to read than it should be...

File: pcdsdevices/spectrometer.py (hunk 1 above)
Comment (ZLLentz) [reply pcdsdevices/spectrometer.py:314]: You are 100% correct

File: pcdsdevices/state.py (hunk 7 above)
Comment (ZLLentz) [reply pcdsdevices/state.py:737]: I need re-read these and double-check on index 0 and on the other one above

Comment (ZLLentz) [reply pcdsdevices/state.py]: These must have been copy/paste errors from the other function that has a None in the list.
I'm going to evaluate that None right now but I've fixed these two type annotations

**Pull request merged**